File: modules/api/src/test/net/countercraft/movecraft/SolidHitBoxTest.java
package net.countercraft.movecraft;

import net.countercraft.movecraft.util.hitboxes.HitBox;
import net.countercraft.movecraft.util.hitboxes.SolidHitBox;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsInAnyOrder;
import static org.hamcrest.Matchers.iterableWithSize;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertIterableEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;


public class SolidHitBoxTest {

    final private MovecraftLocation minimum = new MovecraftLocation(0,0,0);
    final private MovecraftLocation maximum = new MovecraftLocation(3,0,3);
    final private HitBox hitBox = new SolidHitBox(minimum, maximum);

    @Test
    public void testBounds(){
        assertEquals(minimum.getX(), hitBox.getMinX());
        assertEquals(minimum.getY(), hitBox.getMinY());
        assertEquals(minimum.getZ(), hitBox.getMinZ());
        assertEquals(maximum.getX(), hitBox.getMaxX());
        assertEquals(maximum.getY(), hitBox.getMaxY());
        assertEquals(maximum.getZ(), hitBox.getMaxZ());
    }

    @Test
    public void testSize(){
        assertEquals(16, hitBox.size());
        assertThat(hitBox, iterableWithSize(16));
    }


    @Test
    public void testContents(){
        final List<MovecraftLocation> locations = new ArrayList<>(16);
        for(int x = minimum.getX(); x <= maximum.getX(); x++){
            for(int y = minimum.getY(); y <= maximum.getY(); y++){
                for(int z = minimum.getX(); z <= maximum.getZ(); z++){
                    locations.add(new MovecraftLocation(x,y,z));
                }
            }
        }
        assertThat(locations,containsInAnyOrder(hitBox.asSet().toArray()));
    }

    @Test
    public void testBoundingHitBox(){
        assertIterableEquals(hitBox, hitBox.boundingHitBox());
    }

    @Test
    public void testContainsAll(){
        final List<MovecraftLocation> locations = new ArrayList<>(16);
        for(int x = minimum.getX(); x <= maximum.getX(); x++){
            for(int y = minimum.getY(); y <= maximum.getY(); y++){
                for(int z = minimum.getX(); z <= maximum.getY(); z++){
                    locations.add(new MovecraftLocation(x,y,z));
                }
            }
        }
        assertTrue(hitBox.containsAll(locations));
    }



}

File: src/test/kotlin/SolidHitBoxTest.kt
import dev.cccm5.movecraft.hitbox.SolidHitBox
import dev.cccm5.movecraft.util.GridVector
import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import com.natpryce.hamkrest.has
import com.natpryce.hamkrest.isIn
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import kotlin.test.assertEquals


@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class SolidHitBoxTest {
    private fun hasSize(size: Int) = has(Collection<Any>::size, equalTo(size))
    private val minimum = GridVector(0, 0, 0)
    private val maximum = GridVector(3, 3, 3)
    private val hitBox = SolidHitBox(minimum, maximum)

    @Test
    fun testBounds(){
        assertEquals(minimum.x, hitBox.minX)
        assertEquals(minimum.y, hitBox.minY)
        assertEquals(minimum.z, hitBox.minZ)
        assertEquals(maximum.x, hitBox.maxX)
        assertEquals(maximum.y, hitBox.maxY)
        assertEquals(maximum.z, hitBox.maxZ)
    }

    @Test
    fun testSize() {
        assertEquals(64, hitBox.size)
        assertThat(hitBox, hasSize(64))
    }

    @Test
    fun testContents(){
        for (x in minimum.x..maximum.x) {
            for (y in minimum.y..maximum.y) {
                for (z in minimum.x..maximum.y) {
                    assertThat(GridVector(x, y, z), isIn(hitBox))
                }
            }
        }
    }

    @Test
    fun testBoundingHitBox(){
        assertEquals(hitBox, hitBox.boundingHitBox())
    }

    @Test
    fun testContainsAll() {
        val locations = mutableListOf<GridVector>()
        for (x in minimum.x..maximum.x) {
            for (y in minimum.y..maximum.y) {
                for (z in minimum.x..maximum.x) {
                    locations.add(GridVector(x, y, z))
                }
            }
        }
        assert(hitBox.containsAll(locations))
    }
}